src/swapd/runtime.rs

```diff
@@ -771,22 +771,24 @@ impl Runtime {
                             "{} | checkpointing alice pre buy swapd state",
                             self.swap_id.bright_blue_italic()
                         );
-                        checkpoint_send(
-                            endpoints,
-                            self.swap_id,
-                            self.identity(),
-                            ServiceId::Database,
-                            request::CheckpointState::CheckpointSwapd(CheckpointSwapd {
-                                state: self.state.clone(),
-                                last_msg: Msg::BuyProcedureSignature(buy_proc_sig.clone()),
-                                enquirer: self.enquirer.clone(),
-                                temporal_safety: self.temporal_safety.clone(),
-                                txs: self.txs.clone(),
-                                txids: self.syncer_state.tasks.txids.clone(),
-                                pending_requests: self.pending_requests.clone(),
-                                xmr_addr_addendum: self.syncer_state.xmr_addr_addendum.clone(),
-                            }),
-                        )?;
+                        if self.state.a_sup_checkpoint_pre_buy() {
```

Comment:
this one comment extends to the rest of the pr related to state updates, sup. Here one updates the state before being sure that checkpoint_send is successful, therefore generating some potential invalid and unhandled cases

Reply:
Yeah, we need a workflow for critical calls like this where we wait for a response from the database before progressing in the protocol. But I need to think about this some more, so work for another PR.

Reply:
you're right, thats a larger problem